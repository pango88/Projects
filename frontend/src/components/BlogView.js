import React from 'react';
import { addLike, deleteBlog, createComment } from '../reducers/blogReducer';
import { useDispatch, useSelector } from 'react-redux';
import { useHistory } from 'react-router-dom';

const BlogView = ({ blog }) => {
  const blogs = useSelector((state) => state.blogs);
  const dispatch = useDispatch();
  const history = useHistory();

  const boldSpan = {
    fontWeight: 'bold',
  };

  // likar lite segt
  const likeBlog = async (id) => {
    const blog = blogs.find((blog) => blog.id === id);
    dispatch(addLike(id, blog));
  };

  // error handling does not work
  const removeBlog = async (id) => {
    const confirmation = window.confirm(
      'Are you sure you want to delete this blog?'
    );
    if (confirmation) {
      dispatch(deleteBlog(id));
      history.goBack();
    }
  };

  const addComment = (event) => {
    event.preventDefault();
    const newComment = {
      content: event.target.content.value,
    };
    dispatch(createComment(blog.id, newComment));
    event.target.content.value = '';
  };

  if (!blog) {
    return null;
  }

  return (
    <div className="blog">
      {blog.title} <span style={boldSpan}>by</span> {blog.author}{' '}
      <p>
        <span style={boldSpan}>url:</span> {blog.url}
      </p>
      <p id="likes">
        <span style={boldSpan}>likes:</span> {blog.likes}
        <button id="like" onClick={() => likeBlog(blog.id)} className="likeBtn">
          like
        </button>
      </p>
      <p>
        <span style={boldSpan}>creator:</span> {blog.user.name}
      </p>
      <h3>Comments</h3>
      <form onSubmit={addComment}>
        <input id="comment" type="text" name="content" />
        <button type="submit">add comment</button>
      </form>
      <ul>
        {blog.comments.map((comment) => (
          <li key={comment.id}>{comment.content}</li>
        ))}
      </ul>
      <button
        id="#delete"
        onClick={() => {
          removeBlog(blog.id);
        }}
        style={{ color: 'red', fontWeight: 'bold' }}
      >
        remove
      </button>
    </div>
  );
};

export default BlogView;
