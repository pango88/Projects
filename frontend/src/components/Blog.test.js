import React from 'react';
import { render } from '@testing-library/react';
import { useSelector } from 'react-redux';
import BlogView from './BlogView';

const testBlog = {
  title: 'testTitle',
  author: 'reduxTesting',
  url: 'jest.com',
  likes: '0',
  id: '12312312321',
  user: { name: 'whatever' },
  comments: [{ content: 'testComment', id: '123' }],
};

jest.mock('react-redux', () => ({
  useDispatch: jest.fn(),
  useSelector: jest.fn(),
}));

describe('BlogView', () => {
  test('renders blogView correctly', () => {
    // The selector isnt really needed so much but the component uses the blogs to find the correct id for the handlers
    useSelector.mockImplementation((selector) =>
      selector({
        blogs: [
          {
            title: 'testTitle',
            author: 'reduxTesting',
            url: 'jest.com',
            likes: '0',
            id: '12312312321',
            user: { name: 'whatever' },
          },
        ],
      })
    );
    const { container } = render(<BlogView blog={testBlog} />);
    expect(container).toMatchSnapshot();
    expect(container).toHaveTextContent('testTitle');
  });
});
