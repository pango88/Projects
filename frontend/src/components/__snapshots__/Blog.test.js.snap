// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BlogView renders blogView correctly 1`] = `
<div>
  <div
    class="blog"
  >
    testTitle
     
    <span
      style="font-weight: bold;"
    >
      by
    </span>
     
    reduxTesting
     
    <p>
      <span
        style="font-weight: bold;"
      >
        url:
      </span>
       
      jest.com
    </p>
    <p
      id="likes"
    >
      <span
        style="font-weight: bold;"
      >
        likes:
      </span>
       
      0
      <button
        class="likeBtn"
        id="like"
      >
        like
      </button>
    </p>
    <p>
      <span
        style="font-weight: bold;"
      >
        creator:
      </span>
       
      whatever
    </p>
    <h3>
      Comments
    </h3>
    <form>
      <input
        id="comment"
        name="content"
        type="text"
      />
      <button
        type="submit"
      >
        add comment
      </button>
    </form>
    <ul>
      <li>
        testComment
      </li>
    </ul>
    <button
      id="#delete"
      style="color: red; font-weight: bold;"
    >
      remove
    </button>
  </div>
</div>
`;
